lib/MonadicRewrite.thy

```diff
@@ -507,39 +625,27 @@ next
     by (fastforce simp: corres_underlying_def split_def)
 qed
 
-lemma monadic_rewrite_refine_valid:
-  "monadic_rewrite F E P f g
-    \<Longrightarrow> \<lbrace>P'\<rbrace> f \<lbrace>Q\<rbrace>
-    \<Longrightarrow> F \<longrightarrow> no_fail P'' f
-    \<Longrightarrow> \<lbrace>P and P' and P''\<rbrace> g \<lbrace>Q\<rbrace>"
-  apply (clarsimp simp: monadic_rewrite_def valid_def no_fail_def imp_conjL)
-  apply (drule spec, drule(1) mp)+
-  apply (clarsimp simp: Ball_def)
-  apply auto
-  done
+lemma monadic_rewrite_corres_rhs:
+  "\<lbrakk> monadic_rewrite False True Q c c';
+     corres_underlying R F F' r P P' a c' \<rbrakk>
+   \<Longrightarrow> corres_underlying R F F' r P (P' and Q) a c"
+  by (fastforce simp: corres_underlying_def monadic_rewrite_def)
 
-lemma monadic_rewrite_refine_validE_R:
-  "monadic_rewrite F E P f g
-    \<Longrightarrow> \<lbrace>P'\<rbrace> f \<lbrace>Q\<rbrace>, -
-    \<Longrightarrow> F \<longrightarrow> no_fail P'' f
-    \<Longrightarrow> \<lbrace>P and P' and P''\<rbrace> g \<lbrace>Q\<rbrace>, -"
-  by (simp add: validE_R_def validE_def monadic_rewrite_refine_valid)
+lemma monadic_rewrite_corres2:
+  "\<lbrakk> monadic_rewrite False E Q a a';
+     corres_underlying R False F r P P' a' c \<rbrakk>
+   \<Longrightarrow> corres_underlying R False F r (P and Q) P' a c"
```

Comment:
Doesn't have to be as part of this PR, but for consistency with the previous lemma should this be `monadic_rewrite_corres_lhs`?